### PR TITLE
msg: allow different ms type for cluster network and public network

### PR DESCRIPTION
**Pull request by yuyuyu101**

### CODE REVIEW

**Comment by yuyuyu101** [comment]
in order to make cluster network can talk infiniband/rdma, and public network talk posix/tcpip

**Comment by mattbenjamin** [comment]
@yuyuyu101 posix, or tcp?

**Comment by dachary** [comment]
jenkins test this please (no logs)

**Comment by yuyuyu101** [comment]
@mattbenjamin ok

**Comment by yuyuyu101** [comment]
@liewegas do we need this at k? because I want to remove ms_async_transport_type which introduced just now. if we want to integrate transport type into ms type, this should be a good time. otherwise, we need to make compatible with ms async transport type

**Comment by liewegas** [comment]
combining them into a single field sounds best.  i don't have an opinion 
on whether it has to be in kraken or not, though.

**Comment by Adirl** [comment]
running this patch (after some compile issues) looks like we found few other messengers that need to choose sides. (cluster or public)
@yuyuyu101  what do you think ? 

src/ceph_mds.cc:141:  Messenger *msgr = Messenger::create(g_ceph_context, g_conf->ms_type,
src/mgr/DaemonServer.cc:48:  msgr = Messenger::create(g_ceph_context, g_conf->ms_type,
src/msg/Messenger.cc:19:  return Messenger::create(cct, cct->_conf->ms_type, entity_name_t::CLIENT(),
src/ceph_mon.cc:655: Messenger *msgr = Messenger::create(g_ceph_context, g_conf->ms_type,

src/test/messenger/simple_client.cc:106:        messenger = Messenger::create(g_ceph_context, g_conf->ms_type,
src/test/messenger/simple_server.cc:77: messenger = Messenger::create(g_ceph_context, g_conf->ms_type,
src/test/mon/test-mon-msg.cc:81:    msg = Messenger::create(cct, cct->_conf->ms_type, entity_name_t::CLIENT(-1),
src/test/mon/test_mon_workloadgen.cc:363:    messenger.reset(Messenger::create(cct, cct->_conf->ms_type, entity_name_t::OSD(whoami),
src/test/osd/TestOSDScrub.cc:59:  Messenger *ms = Messenger::create(g_ceph_context, g_conf->ms_type,

**Comment by yuyuyu101** [comment]
I'm not sure, by default ms_type should be the public msgr type.

**Comment by Adirl** [comment]
Ok, 
we will do some testing and propose a patch setting them all as public

**Comment by Adirl** [comment]
@danielbo @saritz

**Comment by DanielBar-On** [comment]
@yuyuyu101 
Hey Haomi,

I'm trying to run your code. Everything works if I set *both* ms_public_type and ms_cluster_type to "async+posix" or "async+rdma". 
When I set ms_public_type to "async+posix" and ms_cluster_type to "async+rdma", the monitor will come up but when trying to start the osd with ceph-osd, I get the following failed assert "/Event.cc: 205: FAILED assert(in_thread()) ". 
The assert is called from create_file_event ( Event.cc: In function 'int EventCenter::create_file_event(int, int, EventCallbackRef)' ),

Why do you think the crash is happening?
Le me know if you need more information.

**Comment by yuyuyu101** [comment]
@DanielBar-On do you have call stack dump ?

**Comment by DanielBar-On** [comment]
@yuyuyu101 
/.autodirect/mtrswgwork/danielbo/febe/ceph2/ceph/src/msg/async/Event.cc: In function 'int EventCenter::create_file_event(int, int, EventCallbackRef)' thread 7f81a8cb3700 time 2017-01-12 18:04:03.611833
/.autodirect/mtrswgwork/danielbo/febe/ceph2/ceph/src/msg/async/Event.cc: 205: FAILED assert(in_thread())                                                                                                 
 ceph version 11.1.0-25-g47475c5 (47475c511e1e392b9cf166b5f8152c03c1e5c6a0)                                                                                                                              
 1: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x85) [0x7f81b0c50825]                                                                                                         
 2: (EventCenter::create_file_event(int, int, EventCallback*)+0x82b) [0x7f81b0cdba6b]                                                                                                                    
 3: (()+0xb3e988) [0x7f81b0cd0988]                                                                                                                                                                       
 4: (EventCenter::process_events(int)+0x80c) [0x7f81b0cdc93c]                                                                                                                                            
 5: (()+0xb4d072) [0x7f81b0cdf072]                                                                                                                                                                       
 6: (()+0xb5230) [0x7f81ad35c230]                                                                                                                                                                        
 7: (()+0x7dc5) [0x7f81adbdddc5]                                                                                                                                                                         
 8: (clone()+0x6d) [0x7f81acac3ced]                                                                                                                                                                      
 NOTE: a copy of the executable, or `objdump -rdS <executable>` is needed to interpret this.                                                                                                             
2017-01-12 18:04:03.616250 7f81a8cb3700 -1 /.autodirect/mtrswgwork/danielbo/febe/ceph2/ceph/src/msg/async/Event.cc: In function 'int EventCenter::create_file_event(int, int, EventCallbackRef)' thread 7f81a8cb3700 time 2017-01-12 18:04:03.611833                                                                                                                                                                          
/.autodirect/mtrswgwork/danielbo/febe/ceph2/ceph/src/msg/async/Event.cc: 205: FAILED assert(in_thread())                                                                                                       

 ceph version 11.1.0-25-g47475c5 (47475c511e1e392b9cf166b5f8152c03c1e5c6a0)
 1: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x85) [0x7f81b0c50825]
 2: (EventCenter::create_file_event(int, int, EventCallback*)+0x82b) [0x7f81b0cdba6b]           
 3: (()+0xb3e988) [0x7f81b0cd0988]                                                              
 4: (EventCenter::process_events(int)+0x80c) [0x7f81b0cdc93c]                                   
 5: (()+0xb4d072) [0x7f81b0cdf072]                                                              
 6: (()+0xb5230) [0x7f81ad35c230]                                                               
 7: (()+0x7dc5) [0x7f81adbdddc5]                                                                
 8: (clone()+0x6d) [0x7f81acac3ced]                                                             
 NOTE: a copy of the executable, or `objdump -rdS <executable>` is needed to interpret this.    

     0> 2017-01-12 18:04:03.616250 7f81a8cb3700 -1 /.autodirect/mtrswgwork/danielbo/febe/ceph2/ceph/src/msg/async/Event.cc: In function 'int EventCenter::create_file_event(int, int, EventCallbackRef)' thread 7f81a8cb3700 time 2017-01-12 18:04:03.611833                                                                                                                                                                  
/.autodirect/mtrswgwork/danielbo/febe/ceph2/ceph/src/msg/async/Event.cc: 205: FAILED assert(in_thread())                                                                                                       

 ceph version 11.1.0-25-g47475c5 (47475c511e1e392b9cf166b5f8152c03c1e5c6a0)
 1: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x85) [0x7f81b0c50825]
 2: (EventCenter::create_file_event(int, int, EventCallback*)+0x82b) [0x7f81b0cdba6b]           
 3: (()+0xb3e988) [0x7f81b0cd0988]                                                              
 4: (EventCenter::process_events(int)+0x80c) [0x7f81b0cdc93c]                                   
 5: (()+0xb4d072) [0x7f81b0cdf072]                                                              
 6: (()+0xb5230) [0x7f81ad35c230]                                                               
 7: (()+0x7dc5) [0x7f81adbdddc5]                                                                
 8: (clone()+0x6d) [0x7f81acac3ced]                                                             
 NOTE: a copy of the executable, or `objdump -rdS <executable>` is needed to interpret this.    

*** Caught signal (Aborted) **
 in thread 7f81a8cb3700 thread_name:ceph-osd
 ceph version 11.1.0-25-g47475c5 (47475c511e1e392b9cf166b5f8152c03c1e5c6a0)
 1: (()+0x928f4a) [0x7f81b0abaf4a]                                         
 2: (()+0xf100) [0x7f81adbe5100]                                           
 3: (gsignal()+0x37) [0x7f81aca025f7]
 4: (abort()+0x148) [0x7f81aca03ce8]
 5: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x267) [0x7f81b0c50a07]
 6: (EventCenter::create_file_event(int, int, EventCallback*)+0x82b) [0x7f81b0cdba6b]
 7: (()+0xb3e988) [0x7f81b0cd0988]
 8: (EventCenter::process_events(int)+0x80c) [0x7f81b0cdc93c]
 9: (()+0xb4d072) [0x7f81b0cdf072]
 10: (()+0xb5230) [0x7f81ad35c230]
 11: (()+0x7dc5) [0x7f81adbdddc5]
 12: (clone()+0x6d) [0x7f81acac3ced]
2017-01-12 18:04:03.625228 7f81a8cb3700 -1 *** Caught signal (Aborted) **
 in thread 7f81a8cb3700 thread_name:ceph-osd

 ceph version 11.1.0-25-g47475c5 (47475c511e1e392b9cf166b5f8152c03c1e5c6a0)
 1: (()+0x928f4a) [0x7f81b0abaf4a]
 2: (()+0xf100) [0x7f81adbe5100]
 3: (gsignal()+0x37) [0x7f81aca025f7]
 4: (abort()+0x148) [0x7f81aca03ce8]
 5: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x267) [0x7f81b0c50a07]
 6: (EventCenter::create_file_event(int, int, EventCallback*)+0x82b) [0x7f81b0cdba6b]
 7: (()+0xb3e988) [0x7f81b0cd0988]
 8: (EventCenter::process_events(int)+0x80c) [0x7f81b0cdc93c]
 9: (()+0xb4d072) [0x7f81b0cdf072]
 10: (()+0xb5230) [0x7f81ad35c230]
 11: (()+0x7dc5) [0x7f81adbdddc5]
 12: (clone()+0x6d) [0x7f81acac3ced]
 NOTE: a copy of the executable, or `objdump -rdS <executable>` is needed to interpret this.

     0> 2017-01-12 18:04:03.625228 7f81a8cb3700 -1 *** Caught signal (Aborted) **
 in thread 7f81a8cb3700 thread_name:ceph-osd

 ceph version 11.1.0-25-g47475c5 (47475c511e1e392b9cf166b5f8152c03c1e5c6a0)
 1: (()+0x928f4a) [0x7f81b0abaf4a]
 2: (()+0xf100) [0x7f81adbe5100]
 3: (gsignal()+0x37) [0x7f81aca025f7]
 4: (abort()+0x148) [0x7f81aca03ce8]
 5: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x267) [0x7f81b0c50a07]
 6: (EventCenter::create_file_event(int, int, EventCallback*)+0x82b) [0x7f81b0cdba6b]
 7: (()+0xb3e988) [0x7f81b0cd0988]
 8: (EventCenter::process_events(int)+0x80c) [0x7f81b0cdc93c]
 9: (()+0xb4d072) [0x7f81b0cdf072]
 10: (()+0xb5230) [0x7f81ad35c230]
 11: (()+0x7dc5) [0x7f81adbdddc5]
 12: (clone()+0x6d) [0x7f81acac3ced]
 NOTE: a copy of the executable, or `objdump -rdS <executable>` is needed to interpret this.

File: src/msg/Messenger.cc [hunk 1]
```diff
@@ -36,8 +37,8 @@ Messenger *Messenger::create(CephContext *cct, const string &type,
   }
   if (r == 0 || type == "simple")
     return new SimpleMessenger(cct, name, std::move(lname), nonce);
-  else if (r == 1 || type == "async")
-    return new AsyncMessenger(cct, name, std::move(lname), nonce);
+  else if (r == 1 || type.find("async") != std::string::npos)
```
Comment (liewegas) [inline src/msg/Messenger.cc]: type.find("async") == 0  ?

File: src/msg/async/AsyncMessenger.cc [hunk 2]
```diff
@@ -235,17 +239,24 @@ class C_handle_reap : public EventCallback {
...
+                               const std::string &type, string mname, uint64_t _nonce)
   : SimplePolicyMessenger(cct, name,mname, _nonce),
     dispatch_queue(cct, this, mname),
     lock("AsyncMessenger::lock"),
     nonce(_nonce), need_addr(true), did_bind(false),
     global_seq(0), deleted_lock("AsyncMessenger::deleted_lock"),
     cluster_protocol(0), stopped(true)
 {
+  std::string transport_type = "posix";
+  if (type.find("rdma") != std::string::npos)
```
Comment (liewegas) [inline src/msg/async/AsyncMessenger.cc]: +rdma

File: src/msg/async/AsyncMessenger.cc [hunk 3]
```diff
@@ -235,17 +239,24 @@ class C_handle_reap : public EventCallback {
...
+    transport_type = "rdma";
+  else if (type.find("dpdk") != std::string::npos)
```
Comment (liewegas) [inline src/msg/async/AsyncMessenger.cc]: +dpdk

**Comment by yuyuyu101** [comment]
my idea is want to allow user to specify posix+async.... I don't want to restrict too much on this...

**Comment by liewegas** [comment]
yeah sounds good!

File: src/msg/async/Event.cc [hunk 4]
```diff
@@ -99,14 +99,15 @@ ostream& EventCenter::_event_prefix(std::ostream *_dout)
                 << " time_id=" << time_event_next_id << ").";
 }
 
-int EventCenter::init(int n, unsigned i)
+int EventCenter::init(int n, unsigned i, const std::string &t)
```
Comment (tchaikov) [inline src/msg/async/Event.cc]:
```
CMakeFiles/test_trans.dir/test_trans.cc.o -c /build/ceph-11.1.0-6908-g631f788/src/test/test_trans.cc
/build/ceph-11.1.0-6908-g631f788/src/test/perf_local.cc: In function 'double eventcenter_poll()':
/build/ceph-11.1.0-6908-g631f788/src/test/perf_local.cc:453:22: error: no matching function for call to 'EventCenter::init(int, int)'
   center.init(1000, 0);
                      ^
/build/ceph-11.1.0-6908-g631f788/src/test/perf_local.cc:453:22: note: candidate is:
In file included from /build/ceph-11.1.0-6908-g631f788/src/test/perf_local.cc:55:0:
/build/ceph-11.1.0-6908-g631f788/src/msg/async/Event.h:194:7: note: int EventCenter::init(int, unsigned int, const string&)
   int init(int nevent, unsigned idx, const std::string &t);
       ^
/build/ceph-11.1.0-6908-g631f788/src/msg/async/Event.h:194:7: note:   candidate expects 3 arguments, 2 provided
/build/ceph-11.1.0-6908-g631f788/src/test/perf_local.cc: In constructor 'CenterWorker::CenterWorker(CephContext*)':
/build/ceph-11.1.0-6908-g631f788/src/test/perf_local.cc:470:23: error: no matching function for call to 'EventCenter::init(int, int)'
     center.init(100, 0);
                       ^
/build/ceph-11.1.0-6908-g631f788/src/test/perf_local.cc:470:23: note: candidate is:
In file included from /build/ceph-11.1.0-6908-g631f788/src/test/perf_local.cc:55:0:
/build/ceph-11.1.0-6908-g631f788/src/msg/async/Event.h:194:7: note: int EventCenter::init(int, unsigned int, const string&)
   int init(int nevent, unsigned idx, const std::string &t);
       ^
/build/ceph-11.1.0-6908-g631f788/src/msg/async/Event.h:194:7: note:   candidate expects 3 arguments, 2 provided
make[4]: Leaving directory `/build/ceph-11.1.0-6908-g631f788/obj-x86_64-linux-gnu'
```

see https://jenkins.ceph.com/job/ceph-dev-new-build/ARCH=x86_64,AVAILABLE_ARCH=x86_64,AVAILABLE_DIST=trusty,DIST=trusty,MACHINE_SIZE=huge/638//consoleFull

**Comment by Adirl** [comment]
we have a cluster running this code with RDMA in backend and TCP in frontend. 
health is OK and fio traffic from client looks good

**Comment by tchaikov** [comment]
needs rebase.

**Comment by Adirl** [comment]
@yuyuyu101 
note when rebasing that the latest master is broken for RDMA, 
we have pending PR for that: 
https://github.com/ceph/ceph/pull/13122#issuecomment-275339141

**Comment by Adirl** [comment]
looks like we need this change in ceph_mds.cc +146

  Messenger *msgr = Messenger::create(g_ceph_context, g_conf->ms_type,
				      entity_name_t::MDS(-1), "mds",
				      nonce, Messenger::HAS_MANY_CONNECTIONS);

**Comment by yuriw** [comment]
http://pulpito.front.sepia.ceph.com/yuriw-2017-02-02_16:18:13-rados-wip-yuri-testing_2017_2_1-distro-basic-smithi/

http://pulpito.front.sepia.ceph.com:80/yuriw-2017-02-02_21:52:16-rados-wip-yuri-testing_2017_2_1---basic-smithi/

**Pull request merged**